Feature/sim 1101/photometric parameters

danielsf opened the pull request:
This pull request takes the information previously stored in the PhotometricDefaults class and breaks it into three categories:

PhotometricParameters stores telescope parameters (e.g. effective area, read noise, plate scale).  All of the methods which rely on this data now take as argument a PhotometricParameters instantiation.  This will allow people from other projects to more easily adapt our code to non-LSST systems.

PhysicalParameters stores things like physical constants that we just want to have access to at different places in the stack.

LSSTdefaults (I'm not married to that name) stores the default values of seeing, m5, and the noise gamma paramter from the overview paper.  This is used in cases where the user has not specified an m5 value through ObservationMetaData.

rhiannonlynne commented on python/lsst/sims/photUtils/Sed.py:
```diff
@@ -106,6 +106,10 @@ def __init__(self, wavelen=None, flambda=None, fnu=None, badval=numpy.NaN, name=
         self.badval = badval
+
+        self._physParams = PhysicalParameters()
+        self._seedingDefaults = LSSTdefaults()
```
Is this a typo? (seedingDefaults is not used anywhere else, but _seeingDefaults is. 
Does LSSTdefaults only carry the seeing default values (at least, those are the only parameters relevant for this class?)

danielsf replied:
that is, in fact, a typo.  Thanks.

danielsf replied:
I'm actually going to change this and make _seeingDefaults a dict copied straight from LSSTdefaults, i.e.

`self._seeingDefaults = LSSTdefault()._seeing`

rhiannonlynne commented on python/lsst/sims/photUtils/Sed.py:
```diff
@@ -636,15 +654,32 @@ def multiplySED(self, other_sed, wavelen_step=PhotometricDefaults.wavelenstep):
+        photometric response of the telescope.
         """
+
+        if photParams is None:
```
I don't think this is how I would handle photParams. If it's not optional, don't make it an optional kwarg, make it an argument. 
Or you could make it optional, and have the class itself import the "basic default" photometricParameters class in the init? (given that it's not optional in Bandpass however, I think making it not optional here and an argument to the method, is the right approach).

danielsf replied:
Makes sense.  I was enamored of the idea of not having to remember the order of arguments whenever I call these functions, but I was forgetting that python allows you to assign args using the fn(x=1, y=2, z=3) formalism, as well.

I have 'demoted' photParams from a kwarg to an arg in all of the relevant methods.

rhiannonlynne commented on python/lsst/sims/photUtils/Sed.py:
```diff
@@ -932,28 +971,28 @@ def calcNeff(self, seeing, platescale):
 
-    def calcNonSourceNoiseSq(self, skySed, hardwarebandpass, readnoise, darkcurrent,
-                           othernoise, seeing, effarea, expTime, nexp, platescale, gain):
+        if photParams is None:
```
Again - if it's not optional, it shouldn't be a kwarg.

rhiannonlynne commented on python/lsst/sims/photUtils/Sed.py:
```diff
@@ -995,23 +1022,26 @@ def calcNonSourceNoiseSq(self, skySed, hardwarebandpass, readnoise, darkcurrent,
+
         #Calculate the effective number of pixels for double-Gaussian PSF
-        neff = self.calcNeff(seeing, platescale)
+        neff = self.calcNeff(seeing, photParams.platescale)
```
I don't understand why the seeing doesn't have a default value though (both here and in the Neff method call itself), given that you set the _seeingDefault values in the init. It seems inconsistent - either don't set it in the init, or use those defaults when calculating neff. Same with the platescale in calcNeff method. 
(I guess what I am arguing for is to make the seeing and platescale arguments in calcNeff to be kwargs, set to the class's default seeing and platescale values, and then let the user override when calling the various calc*Noise and calcSNR_PSF methods, but that those would also have the default values).

danielsf replied:
I think I'd rather just remove the default seeing values.  I am worried about users forgetting to set the seeing kwarg and then accidentally using LSST default values when they did not mean to.

rhiannonlynne commented on python/lsst/sims/photUtils/Sed.py:
```diff
@@ -995,23 +1022,26 @@ def calcNonSourceNoiseSq(self, skySed, hardwarebandpass, readnoise, darkcurrent,
         #This method outputs all of the parameters calculated along the way
         #so that the verbose version of calcSNR_psf still works
 
+        if photParams is None:
```
again - if there's no reasonable default, it shouldn't be a kwarg.

rhiannonlynne commented:
Given the improvements you have made in the 'noiseUtilities' class, I am wondering if some of the methods in Sed should still be in Sed. For example, having calcSNR_psf here and calcSNR_gamma in noiseUtilities seems inconsistent to me. 
However, even if the various SNR and noise methods remain there (and I think I prefer this in some ways, mostly because of existing code, although logically I would then want the noiseUtilities code to go back into Sed and Bandpass too), then I think perhaps 'calcSNR_mag' should go. 
You have a better method that includes the gamma correction in noiseUtilities, and I'd actually suggest moving that method from noiseUtilities into Sed to replace calcSNR_mag with calcSNR_gamma. 
You *could* make it easier/simpler by pre-calculating the default gammas in the LSSTDefaults class (along with the default seeing, default m5 values, you could have default gamma), and then just using a representative gamma here (pick 'r' band?). The differences would be minor. 
Another option would be to leave the calcSNR_gamma in the noiseUtilities class, and just update the calcSNR_mag here with those default gamma parameters (or let the user override gamma), with appropriate comments that it's a shortcut version of the one in noiseUtilities that only includes the default gammas. 

I'm curious - how much different is the SNR returned by calcSNR_gamma from calcSNR_PSF? (I think they should be identical once you go to the level of detail you have in calcSNR_gamma).

rhiannonlynne commented:
There's a couple of places you use 'skyBandpass' which is really confusing, especially because there is the similar-sounding 'skySed'. I would recommend using 'atmoBandpass' or 'atmosphere' or something with 'atmosphere' sounding link and referring to it as atmospheric transmission curve, for consistency with existing use throughout calibration and system engineering.

rhiannonlynne commented:
Every time I read 'sig2sys' I think it's something converting sig to sys (and then wonder what those are and what they do). The comment 'square of the systematic noise associated with a particular catalog/measurement' is useful. If this isn't already deeply embedded in other code or usage, would it be possible to change it to sigmaSysSq or something along those lines? 
Also, are default values of these in the LSSTDefaults file? (seems like it would be useful so that people do not exceed our expected noise floor).

rhiannonlynne commented:
I've been trying to watch all the API changes, but I'm not sure I caught them all. 
How much external code do you think needs to be updated to use the new versions of the classes? 
In particular I'm thinking things like 
http://nbviewer.ipython.org/github/LSST/throughputs/blob/master/examples/Calculating%20Magnitudes.ipynb
and 
https://github.com/uwssg/LSST-Tutorials/blob/master/SE/Calculating%20SNR.ipynb
(which is easy enough to update, but we will have to drag a lot of camera team / SE people along with us, with code that is not in git.)

danielsf commented:
calcSNR_gamma and calcSNR_psf agree to within 1 part in 10^3.  At least, now they do.  There was a bug in that calcSNR_gamma was ignoring the zeropoint when converting between magnitudes and fluxes.  I have fixed that by adding methods magnitudeFromFlux and fluxFromMagnitude to the Sed class (which is where we are storing the zeropoint value).

I can see this as an argument in favor of your motion to move all of the noise methods back into the Sed class.  Originally, I moved them out because a lot of them rely on instantiating a flat Sed and it seemed odd to me to have a member method of Sed that constructs a brand new Sed.  I can't think of an actual reason that could get us into trouble, it just felt odd.

One question I am wrestling with: are we sure that the zeropoint should be stored in the Sed class.  Is it something that we want to allow users to change (if they wanted to use a different magnitude system)?  Or should we make it immutable?

You asked about default gamma values.  There are default gamma values stored in LSSTdefaults.  I will open a ticket to refactor the noise calculation routines and we can sit down and actually have a design discussion about what methods to keep, what methods to boot, and where to put them all.

As far as changing the non-Git code that relies on the noise calculation methods, the examples you showed me won't require much change.  There are just one or two places where an instance of PhotometricParameters needs to be passed.  I'll make that change once you okay this for merging.

Let me know if there is anything else you want done before I merge this ticket.

rhiannonlynne commented:
About the zeropoint - it is nice to keep it with the Sed class, so that the user can convert between flux and magnitudes, and it also scales with the values in the Sed class (and is used by fluxNorm, etc). 
So, I think it belongs here. 
As to whether to make it immutable or otherwise more obvious when it is set for AB magnitudes .. I could be swayed either way. LSST is officially AB mags, so I'd lean towards making it immutable. 
You can always compensate for this (in magnitude world) outside Sed/Bandpass themselves (I've done this when converting to Vega magnitudes).

rhiannonlynne commented:
One last nitpick - make the order of 'photParams, seeing' consistent between the calls in 
calcNonSourceNoiseSq(self, skySed, hardwarebandpass, seeing,
+                             photParams):
and 
calcSNR_psf

(I'd suggest making seeing the second argument of the pair, as you more often pass photParams than seeing). 

Then, good to go!
